pkg/morph/client/multi.go

```diff
@@ -1,102 +1,157 @@
-	})
-	if err != nil {
-		return nil, err
+// next returns the next endpoint and its index
+// to try to connect to.
+// Returns -1 index if there is no known RPC endpoints.
+func (e *endpoints) next() (string, int) {
+	if len(e.list) == 0 {
+		return "", -1
 	}
 
-	err = cli.Init() // magic number is set there based on RPC node answer
-	if err != nil {
-		return nil, err
+	next := e.curr + 1
+	if next == len(e.list) {
+		next = 0
 	}
 
-	sCli := blankSingleClient(cli, x.account, &x.cfg)
-	sCli.notary = x.sharedNotary
+	e.curr = next
 
-	c := &Client{
-		cache:        x.sharedCache,
-		singleClient: sCli,
-	}
+	return e.list[next], next
+}
 
-	x.clients[addr] = c
+// current returns an endpoint and its index the Client
+// is connected to.
+// Returns -1 index if there is no known RPC endpoints
+func (e *endpoints) current() (string, int) {
+	if len(e.list) == 0 {
+		return "", -1
+	}
 
-	return c, nil
+	return e.list[e.curr], e.curr
 }
 
-func (x *multiClient) iterateClients(f func(*Client) error) error {
-	var (
-		firstErr error
-		err      error
-	)
+func (c *Client) switchRPC() bool {
+	c.switchLock.Lock()
+	defer c.switchLock.Unlock()
 
-	for i := range x.endpoints {
-		select {
-		case <-x.cfg.ctx.Done():
-			return x.cfg.ctx.Err()
-		default:
-		}
+	c.client.Close()
 
-		x.clientsMtx.Lock()
+	_, currEndpointIndex := c.endpoints.current()
+	if currEndpointIndex == -1 {
+		// there are no known RPC endpoints to try
+		// to connect to => do not switch
+		return false
+	}
 
-		c, cached := x.clients[x.endpoints[i]]
-		if !cached {
-			c, err = x.createForAddress(x.endpoints[i])
+	for {
+		newEndpoint, index := c.endpoints.next()
+		if index == currEndpointIndex {
+			// all the endpoint have been tried
+			// for connection unsuccessfully
+			return false
 		}
 
-		x.clientsMtx.Unlock()
-
-		if !cached && err != nil {
-			x.cfg.logger.Error("could not open morph client connection",
-				zap.String("endpoint", x.endpoints[i]),
-				zap.String("err", err.Error()),
+		cli, err := newWSClient(c.cfg, newEndpoint)
+		if err != nil {
+			c.logger.Warn("could not establish connection to the switched RPC node",
+				zap.String("endpoint", newEndpoint),
+				zap.Error(err),
 			)
-		} else {
-			err = f(c)
-		}
 
-		if err == nil {
-			return nil
+			continue
 		}
 
-		// we dont need to continue the process after the logical error was encountered
-		if errNeoFS := unwrapNeoFSError(err); errNeoFS != nil {
-			return errNeoFS
+		err = cli.Init()
+		if err != nil {
+			cli.Close()
+			c.logger.Warn("could not init the switched RPC node",
+				zap.String("endpoint", newEndpoint),
+				zap.Error(err),
+			)
+
+			continue
 		}
 
-		// set first error once
-		if firstErr == nil {
-			firstErr = err
+		c.client = cli
+
+		return true
+	}
+}
+
+func (c *Client) notificationLoop() {
+	for {
+		select {
+		case <-c.cfg.ctx.Done():
+			_ = c.UnsubscribeAll()
+			c.close()
+
+			return
+		case <-c.closeChan:
+			_ = c.UnsubscribeAll()
```

Comment:
that is internal channel, not neo-go (neo-go is `case n, ok := <-c.client.Notifications:` below)

it only receives `struct{}` when `Close` has been called, so it is a normal shutdown, not because of a connection lost